lir/algorithms/bootstraps.py

```diff
@@ -0,0 +1,174 @@
+        return FeatureData: The feature data to use for interval estimation.
+        """
+        pass
+
+    def transform(self, instances: FeatureData) -> LLRData:
+        """Transform the provided instances to include the best estimate and confidence intervals.
+
+        param instances: FeatureData: The feature data to transform.
+        return LLRData: The transformed feature data with best estimate and confidence intervals.
+        """
+        if self.f_delta_interval_lower is None or self.f_delta_interval_upper is None:
+            raise ValueError('Bootstrap intervals have not been computed. Please fit the bootstrap first.')
+
+        best_estimate = super().transform(instances)
+        best_1d_estimate = best_estimate.features.reshape(-1)
+        interval_lower = best_1d_estimate + self.f_delta_interval_lower(best_1d_estimate)
+        interval_upper = best_1d_estimate + self.f_delta_interval_upper(best_1d_estimate)
+
+        return best_estimate.replace_as(
+            LLRData, features=np.stack([best_1d_estimate, interval_lower, interval_upper], axis=1)
+        )
+
+    def fit_transform(self, instances: FeatureData) -> LLRData:
+        """Combine fitting and transforming in one step.
+
+        param instances: FeatureData: The feature data to fit and transform.
+        return LLRData: The transformed feature data with best estimate and confidence intervals.
+        """
+        return self.fit(instances).transform(instances)
+
+    def fit(self, instances: FeatureData) -> Self:
+        """Fit the bootstrap system to the provided instances.
+
+        param instances: FeatureData: The feature data to fit the bootstrap system on.
+        return Self: The fitted bootstrap system.
+        """
+
+        all_vals = []
+        rng = np.random.default_rng(self.seed)
+
+        bootstrap_data = self.get_bootstrap_data(instances)
+
+        for _ in range(self.n_bootstraps):
+            sample_index = rng.choice(len(instances), size=len(instances))
+            samples = instances[sample_index]
+            super().fit(samples)
+            all_vals.append(super().transform(bootstrap_data).features.reshape(-1))
+
+        all_vals = np.stack(all_vals, axis=1)
+
+        intervals = np.quantile(all_vals, self.interval, axis=1)
+
+        super().fit(instances)
+        best_estimate_bootstrap_data = super().transform(bootstrap_data)
+        x_vals = best_estimate_bootstrap_data.features.reshape(-1)
+
+        lower = intervals[0] - x_vals
+        upper = intervals[1] - x_vals
+
+        # Use the numeric 1D feature array as x for interpolation (not the FeatureData object).
+        self.f_delta_interval_lower = interp1d(x_vals, lower, bounds_error=False, fill_value=(lower[0], lower[-1]))
+        self.f_delta_interval_upper = interp1d(x_vals, upper, bounds_error=False, fill_value=(upper[0], upper[-1]))
+        return self
+
+
+class BootstrapAtData(Bootstrap):
```

Comment:
Veel overzichtelijker zo. Nu is ook meteen duidelijk waar het verschil zit tussen de twee bootstrap implementaties!